nca/Utils/ExplTracker.py

```diff
@@ -0,0 +1,426 @@
+#
+# Copyright 2022 - IBM Inc. All rights reserved
+# SPDX-License-Identifier: Apache2.0
+#
+
+from nca.Utils.Utils import Singleton
+from nca.Utils.NcaLogger import NcaLogger
+from nca.CoreDS.Peer import PeerSet
+from bs4 import BeautifulSoup
+from bs4.element import Tag
+from nca.CoreDS.ConnectivityProperties import ConnectivityProperties
+
+
+class ExplTracker(metaclass=Singleton):
+    """
+    The Explainability Tracker is used for tracking the elements and their configuration
+    so it will be able to specify which configurations are responsible for each peer and each connection
+    or lack of connection between them.
+
+    The ExplTracker is Singletone
+    Members:
+            ExplDescriptorContainer - A container for all expl' items' in the system.
+            Each entry has a peer or a policy with their name, file and line number of the configurations.
+
+            ExplPeerToPolicyContainer - A container for finding the affecting policies from peers in egress and ingress
+            For each peer name it has a ExplPolicies class object with 3 items:
+                    all_policies - a set of all the policies affecting the current peer
+                    egress_dst - a dict of destination peers allowed in the current peer's egress and for each of them,
+                                 the policies that allows it
+                    ingress_src - a dict of source peers allowed in the current peer's ingress and for each of them,
+                                 the policies that allows it
+            That way, when given a src and dst peers we can extract which policies allow the connection in each side.
+            When there is no connection, we list all the policies that affect the peers, so the user may have all the info
+            to find problems.
+
+            _is_active - flag for checking if expl' was activated
+
+            all_conns - all the calculated connection. This is used to shortcut the check if 2 peers are connected or not
+
+            all_peers - all the peers. This is used for the explain-all feature.
+
+            ep - endpoints configurations (use either full_name for pod mode, or workload_name for deployments mode)
+    """
+
+    def __init__(self, ep=''):
+        self.ExplDescriptorContainer = {}
+        self.ExplPeerToPolicyContainer = {}
+        self._is_active = False
+        self.all_conns = None
+        self.all_peers = None
+        self.ep = ep
+
+        self.add_item('', 0, 'Default-Policy')
+
+    class ExplPolicies:
+        """
+        ExplPolicies holds the policies affecting peers in relation to other peers.
+        That is, for each peer it holds all the peers in its egress and ingress and the policies
+        that has effect on the connection to that peers.
+        """
+
+        def __init__(self):
+            self.egress_dst = {}
+            self.ingress_src = {}
```

Comment:
init as `dict()` and document in a comment as a `map from ... to ...`

Reply:
Done